.config/gitpod/Dockerfile

```diff
@@ -0,0 +1,100 @@
+###################
+#     builder     #
+###################
+# First stage: Install build-time dependencies
+FROM ubuntu:jammy as builder
+
+# hadolint ignore=DL3002
+USER root
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# Copy Python version config file
+COPY .python-version /tmp/
+
+# Update system and install packages
+# hadolint ignore=DL3008,DL3009,DL3013
+RUN PYTHON_VERSION=$(cut -d '.' -f 1,2 /tmp/.python-version) \
+    && apt-get update \
+    && apt-get upgrade -y \
+    && apt-get install -y --no-install-recommends \
+    bash-completion \
+    ca-certificates \
+    curl \
+    git \
+    git-lfs \
+    gnupg \
+    htop \
+    iproute2 \
+    lsb-release \
+    make \
+    nano \
+    python3-pip \
+    "python${PYTHON_VERSION}" \
+    "python${PYTHON_VERSION}-venv" \
+    sudo \
+    tree \
+    vim \
+    wget \
+    && python3 -m pip install --no-cache-dir --upgrade pip \
+    && python3 -m pip install --no-cache-dir virtualenv
+
+############################
+#     docker-installer     #
+############################
+FROM builder as docker-installer
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# hadolint ignore=DL3008,DL3009
+RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg \
+    && echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
+    $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list > /dev/null \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends docker-ce docker-ce-cli containerd.io
+
+###########################
+#      npm-installer      #
+###########################
+FROM docker-installer as npm-installer
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# Set a variable for packages version
+ARG NPM_VERSION=9.7.1
+ARG UNGIT_VERSION=1.5.23
+ARG MEGA_LINTER_RUNNER_VERSION=7.0.0
+ARG MARKDOWN_TABLE_FORMATTER_VERSION=1.4.0
+
+# hadolint ignore=DL3008
+RUN curl -fsSL https://deb.nodesource.com/setup_20.x | bash - \
+    && apt-get install  -y --no-install-recommends nodejs \
+    && npm install -g "npm@${NPM_VERSION}" \
+    && npm install -g \
+    "ungit@${UNGIT_VERSION}" \
+    "mega-linter-runner@${MEGA_LINTER_RUNNER_VERSION}" \
+    "markdown-table-formatter@${MARKDOWN_TABLE_FORMATTER_VERSION}"
+
+###########################
+#         cleanup         #
+###########################
+FROM npm-installer as cleanup
+
+RUN apt-get clean && rm -rf /var/lib/apt/lists/* tmp/* 
+
```

Comment:
This has a conceptual error. (We kinda became Dockerfile specialists with time).
Even though we might think that since it a separate stage, and that we often use that pattern to build optimized images, it doesn't do exactly what you might think.

Since it is a "FROM", then this stage will create a new image by using `cleanup` as a base image, like if it was a published image, but in this case, it is defined in the Dockerfile. Docker images work essentially by applying layers on top of each other, each layer being immutable. Just like if each layer was a tar file, and extracted in the same folder. (It is really like that it works).

So, just before the RUN apt-get clean from line 83, the layers of the image will be all the layers that `cleanup` consists of, and at line 83, we are adding a new layer, that will, once applied, delete files in the filesystem of the image.

That means that what will be downloaded and extracted will contain the files, and later on when finishing pulling, we will remove them. So there are no space gains at all, since they had to be included in the layer like at lines 17-20, 53, and 71, but don't appear in the final state. It is "wasted space". The solution for this is the (ugly) pattern that you often see in Dockerfiles, that you chain the clean commands directly in the same "RUN" as you created them. Each instruction in a Dockerfile makes a layer, that won't change.


Edit: I read a little too fast. Finally you used a `COPY --from` at the last stage that was based off the original one.  So it doesn't really matter inside the final image, since all the stages will be discarded. But I'm not sure that the user added applied on line 91 will really work well if you copy and replace the whole filesystem. If you are to base your image on `ubuntu:jammy`, and then completely replace the whole filesystem with your contents, why need to have the final image download two times the contents of ubuntu:jammy? Once with the FROM (the 6 layers of the base image), and another one with the `COPY --from`, that includes the changes.

If the current final image works as expected, (replacing the whole root filesystem), then changing the base image for the final stage to `scratch`, ie: `FROM scratch`, that will work the same. 
`FROM scratch` means to base the image on nothing, your image won't have layers before it.